src/signature.c

```diff
@@ -24,198 +24,386 @@
-}
+	/* Add the certificates to be verified to the store */
+	if (!(lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file()))) {
+		goto error;
+	}
 
-#else // IMPL == IMPL_FAKE
+	/*  Load the our Root cert, which can be in either DER or PEM format */
+	if (!X509_load_cert_file(lookup, CERTNAME, X509_FILETYPE_PEM)) {
+		goto error;
+	}
 
-bool signature_initialize(const char UNUSED_PARAM *ca_cert_filename)
-{
+	if (crl) {
+		if (!(lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file())) ||
+		   (X509_load_crl_file(lookup, crl, X509_FILETYPE_PEM) != 1)) {
+			goto error;
+		}
+		/* set the flags of the store so that CLRs are consulted */
+		X509_STORE_set_flags(store, X509_V_FLAG_CRL_CHECK | X509_V_FLAG_CRL_CHECK_ALL);
+	}
+
+	/* create a verification context and initialize it */
+	if (!(verify_ctx = X509_STORE_CTX_new())) {
+		goto error;
+	}
+
+	if (X509_STORE_CTX_init(verify_ctx, store, cert, NULL) != 1) {
+		goto error;
+	}
+
+	/* Specify which cert to validate in the verify context.
+	 * This is required because we may add multiple certs to the X509 store,
+	 * but we want to validate a specific one out of the group/chain. */
+	X509_STORE_CTX_set_cert(verify_ctx, cert);
+
+	/* verify the certificate */
+	if (X509_verify_cert(verify_ctx) != 1) {
+		goto error;
+	}
+	/* Certificate verified correctly */
 	return true;
-}
 
-void signature_terminate(void)
-{
+error:
+	ERR_print_errors_fp(stderr);
+
+	if (fp) {
+		fclose(fp);
+	}
+	if (store) {
+		X509_STORE_free(store);
+	}
+	if (lookup) {
+		X509_LOOKUP_free(lookup);
+	}
+	return false;
 }
 
-bool signature_verify(const char UNUSED_PARAM *data_filename, const char UNUSED_PARAM *sig_filename)
+int verify_callback(int ok, X509_STORE_CTX *stor)
 {
-	return true;
+	if (!ok) {
+		fprintf(stderr, "Error: %s\n",
+		X509_verify_cert_error_string(stor->error));
+	}
+	return ok;
 }
 
-bool signature_download_and_verify(const char UNUSED_PARAM *data_url, const char UNUSED_PARAM *data_filename)
+/* Downloads the corresponding signature filename from the
+ * swupd server.
+ *
+ * returns: true if signature was downloaded, false otherwise
+ */
+bool download_and_verify_signature(const char *data_url, const char *data_filename)
 {
-	return true;
+	char *sig_url;
+	char *sig_filename;
+	int ret;
+	bool result;
+
+	string_or_die(&sig_url, "%s.sig", data_url);
+
+	string_or_die(&sig_filename, "%s.sig", data_filename);
+
+	ret = swupd_curl_get_file(sig_url, sig_filename, NULL, NULL, false);
+	if (ret) {
+		result = false;
+	} else {
+		result = verify_signature(data_filename, sig_filename);
+	}
+	free(sig_filename);
+	free(sig_url);
+	return result;
 }
 
-void signature_delete(const char UNUSED_PARAM *data_filename)
+/* Delete the signature file downloaded with download_and_verify_signature() */
+void delete_signature(const char *data_filename)
 {
+	char *sig_filename;
+	struct stat st;
+
+	string_or_die(&sig_filename, "%s.sig", data_filename);
+	if (stat(sig_filename, &st) == 0) {
+		unlink(sig_filename);
```

Comment:
Why delete the sig file here? It makes it difficult to investigate the cause of a failure, if there is one, so unless there is a good reason to remove this, I'd probably leave it